src/vm/analysis/trait_checker/pre_type_check.rs

```diff
@@ -0,0 +1,176 @@
+                Some(x) => x,
+                None => continue 
+            };
+            match define_type {
+                DefineFunctions::Trait => {
+                    if let Some(trait_name) = args[0].match_atom() {
+                        defined_traits.insert(trait_name.clone(), exp.clone());
+                        // Traverse and probe for generics nested in the trait definition
+                        if let Some(trait_definition) = &args[1].match_list() {
+                            self.probe_for_generics(trait_definition, &mut referenced_traits, true)?;
+                        }    
+                    }
+                },
+                DefineFunctions::UseTrait => {
+                    if let Some(trait_name) = args[0].match_atom() {
+                        imported_traits.insert(trait_name.clone(), exp.clone());
+                    }
+                },
+                DefineFunctions::ImplTrait => {
+                    if let Some(trait_identifier) = args[0].match_field() {
+                        implemented_traits.insert(trait_identifier.clone());
+                    }
+                },
+                DefineFunctions::PublicFunction | DefineFunctions::PrivateFunction | DefineFunctions::ReadOnlyFunction => {
+                    // Traverse and probe for generics in functions type definitions
+                    self.probe_for_generics(&args, &mut referenced_traits, true)?;
+                },
+                DefineFunctions::Constant | DefineFunctions::Map | DefineFunctions::PersistedVariable | 
+                DefineFunctions::FungibleToken | DefineFunctions::NonFungibleToken => {
+                    self.probe_for_generics(&args[1..], &mut referenced_traits, false)?;
+                }
+            };
+        }
+
+        let mut orphan_trait_references = HashMap::new();
+        for (trait_name, expr) in &referenced_traits {
+            if !imported_traits.contains_key(trait_name) && !defined_traits.contains_key(trait_name) {
+                orphan_trait_references.insert(trait_name.clone(), expr.clone());
+            }
+        }
+
+        Ok(TraitUsages {
+            defined_traits,
+            imported_traits,
+            referenced_traits,
+            implemented_traits,
+            orphan_trait_references,        
+        })
+    }
+
+    fn probe_for_generics(&mut self, exprs: &[SymbolicExpression], 
+                          referenced_traits: &mut HashMap<ClarityName, SymbolicExpression>, 
+                          should_reference: bool) -> CheckResult<()>  {
+        for expression in exprs.iter() {
+
+            match &expression.expr {
+                List(list) => { self.probe_for_generics(&list, referenced_traits, should_reference)?; },
+                TraitReference(trait_name) => {
+                    if should_reference {
+                        referenced_traits.insert(trait_name.clone(), expression.clone());
```

Comment:
Is it possible to introduce a reference cycle?  Also, what happens if there are multiple traits with the same name defined while you're running this analysis pass?

Reply:
> Is it possible to introduce a reference cycle?

I don't believe so and I have some case tests for that, but feel free to find an issue :)

> Multiple traits with the same name defined.

There are some checks for detecting collisions between imported traits and defined traits, but not between defined traits themselves or imported traits. Addressed with https://github.com/blockstack/blockstack-core/pull/1246/commits/01400b7c3808ba358c1861b9fd6f0b4f9bcbb48d